lib/charms/tls_certificates_interface/v3/tls_certificates.py

```diff
@@ -1885,3 +1905,42 @@ def _find_certificate_in_relation_data(self, csr: str) -> Optional[ProviderCerti
                 continue
             return provider_certificate
         return None
+
+    def _calculate_expiry_notification_time(
+        self,
+        expiry_time: datetime,
+        provider_recommended_notification_time: Optional[int],
+    ) -> datetime:
+        """Calculate a reasonable time to notify the user about the certificate expiry.
+
+        It takes into account the time recommended by the provider and by the requirer.
+        Time recommended by the provider is preferred,
+        then time recommended by the requirer,
+        then dynmaicaly calculated time.
+
+        Args:
+            expiry_time: Certificate expiry time
+            provider_recommended_notification_time:
+                Time in hours prior to expiry to notify the user.
+                Recommended by the provider.
+
+        Returns:
+            datetime: Time to notify the user about the certificate expiry.
+        """
+        if provider_recommended_notification_time:
+            provider_recommendation_time_delta = (
+                expiry_time - timedelta(hours=provider_recommended_notification_time)
+            )
+            if datetime.now(timezone.utc) < provider_recommendation_time_delta:
+                return provider_recommendation_time_delta
+
+        requirer_recommendation_time_delta = (
+            expiry_time - timedelta(hours=self.expiry_notification_time)
+        )
+        if datetime.now(timezone.utc) < requirer_recommendation_time_delta:
+            return requirer_recommendation_time_delta
+
+        calculated_hours = round(
+            (expiry_time - datetime.now(timezone.utc)).total_seconds() / (3600 * 3)
```

Comment:
I think the code is intended to use one-third of the certificate's valid time as the recommended value? Maybe it should subtract the not valid before time (`certificate_object.not_valid_before_utc`) instead of the current time (`datetime.now`).

And, I think `math.ceil` might be a better choice than `round` in cases of extremely short validity times (< 3 hours).